File: library/src/main/java/org/meganekkovr/LookDetector.java
package org.meganekkovr;

import org.joml.Vector3f;

public class LookDetector {

    private static LookDetector instance;
    private final long appPtr;

    private LookDetector(long appPtr) {
        this.appPtr = appPtr;
    }

    public synchronized static void init(long appPtr) {
        if (instance != null)
            throw new IllegalStateException("init was called twice!");

        instance = new LookDetector(appPtr);
    }

    public static LookDetector getInstance() {
        return instance;
    }

    private static native boolean isLookingAt(long appPtr, long entityPtr, long geometryComponentPtr, float[] first, float[] second, boolean axisInWorld);

    public boolean isLookingAt(Entity entity) {
        return isLookingAt(entity, null, null, false);
    }

    public boolean isLookingAt(Entity entity, float[] firstIntersect, float[] secondIntersect, boolean axisInWorld) {

        if (!entity.isShown() || !entity.hasComponent(GeometryComponent.class)) {
            return false;
        }

        GeometryComponent geometry = entity.getComponent(GeometryComponent.class);
        return isLookingAt(appPtr, entity.getNativePointer(), geometry.getNativePointer(), firstIntersect, secondIntersect, axisInWorld);
    }

    public Vector3f getLookingPoint(Entity entity) {
        return getLookingPoint(entity, false);
    }

    public Vector3f getLookingPoint(Entity entity, boolean axisInWorkd) {
        float[] first = new float[3];
        boolean looking = isLookingAt(entity, first, null, axisInWorkd);
        if (looking) {
            return new Vector3f(first[0], first[1], first[2]);
        } else {
            return null;
        }
    }
}

File: library/src/main/java/org/meganekkovr/LookDetector.kt
package org.meganekkovr

import org.joml.Vector3f

class LookDetector private constructor(private val appPtr: Long) {

    @JvmOverloads
    fun isLookingAt(entity: Entity, firstIntersect: FloatArray? = null, secondIntersect: FloatArray? = null, axisInWorld: Boolean = false): Boolean {

        if (!entity.isShown) {
            return false
        }

        val geometry = entity.getComponent<GeometryComponent>() ?: return false
        return isLookingAt(appPtr, entity.nativePointer, geometry.nativePointer, firstIntersect, secondIntersect, axisInWorld)
    }

    @JvmOverloads
    fun getLookingPoint(entity: Entity, axisInWorkd: Boolean = false): Vector3f? {
        val first = FloatArray(3)
        val looking = isLookingAt(entity, first, null, axisInWorkd)
        return if (looking) {
            Vector3f(first[0], first[1], first[2])
        } else {
            null
        }
    }

    private external fun isLookingAt(appPtr: Long, entityPtr: Long, geometryComponentPtr: Long, first: FloatArray?, second: FloatArray?, axisInWorld: Boolean): Boolean

    companion object {

        @JvmStatic
        lateinit var instance: LookDetector

        @Synchronized
        @JvmStatic
        internal fun init(appPtr: Long) {
            instance = LookDetector(appPtr)
        }
    }
}
